### PR TITLE
Allow dropdowns to be opened by MouseDown event

### PR DESCRIPTION
**Pull request by jarvis394**
See https://github.com/ppy/osu-framework/discussions/6584

Adds `ToggleOnMouseDown` boolean property for Dropdown and DropdownHeader.

The flag will use `MouseDownEvent` instead of `ClickEvent` for toggling the menu. Dropdown handles `MouseUpEvent`, looking for active hovers and selecting preselected items if the menu is hovered. Dropdown should close if cursor was released outside the menu and should stay open if it is released on the DropdownHeader.

[osu!_0wpYGXF2M1.webm](https://github.com/user-attachments/assets/c267e9d0-13aa-47b5-8268-51e0fcba7653)

### CODE REVIEW

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs [hunk 1]
```diff
@@ -30,6 +30,11 @@ public bool AlwaysShowSearchBar
             set => SearchBar.AlwaysDisplayOnFocus = value;
         }
 
+        /// <summary>
+        /// Whether parent dropdown <see cref="Dropdown{T}"/> should open/close on OnMouseDown event.
+        /// </summary>
+        public bool ToggleOnMouseDown { get; set; }
```
Comment (bdach) [inline osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: does this need to be a toggleable flag?

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs [hunk 2]
```diff
@@ -135,19 +140,68 @@ private void updateState()
...
+        private bool onClick(ClickEvent e)
+        {
+            // No need to handle dropdown as with this flag it has already been toggled by `onMouseDown` handler
+            if (ToggleOnMouseDown)
+            {
+                // Without this check, when dropdown is opened by clicking outside `SearchBar`,
+                // focus will be lost on `onClick` event -- therefore, closing dropdown.
+                // We need to prevent that by manually focusing on the `SearchBar.textBox`.
+                if (dropdown.MenuState == MenuState.Open)
+                    dropdown.ChangeFocus(SearchBar.Child);
```
Comment (bdach) [inline osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: this access of `.Child()` is turbo dodgy, should be replaced by a `DropdownSearchBar.TakeFocus()` method or something that ensures that the child in question here is actually the textbox that you want it to be

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs [hunk 3]
```diff
@@ -135,19 +140,68 @@ private void updateState()
...
+        /// Handles clicks on the header to open/close the menu.
+        /// </summary>
+        private bool onClick(ClickEvent e)
+        {
+            // No need to handle dropdown as with this flag it has already been toggled by `onMouseDown` handler
+            if (ToggleOnMouseDown)
+            {
+                // Without this check, when dropdown is opened by clicking outside `SearchBar`,
+                // focus will be lost on `onClick` event -- therefore, closing dropdown.
+                // We need to prevent that by manually focusing on the `SearchBar.textBox`.
```
Comment (bdach) [inline osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: would like test coverage for this scenario

File: osu.Framework/Graphics/UserInterface/Dropdown.cs [hunk 4]
```diff
@@ -340,6 +349,27 @@ protected override bool OnKeyDown(KeyDownEvent e)
...
+                // Do not close the menu if we are releasing on the DropdownHeader
+                if (!Header.IsHovered)
+                    Menu.Close();
+                return;
+            }
+
+            // Cursor is inside the menu and possibly selecting an item,
+            // commit that selection and close the menu
+            ((IDropdown)this).CommitPreselection();
+            Menu.Close();
```
Comment (bdach) [inline osu.Framework/Graphics/UserInterface/Dropdown.cs]: this logic is very tangled, would rather have

```diff
diff --git a/osu.Framework/Graphics/UserInterface/Dropdown.cs b/osu.Framework/Graphics/UserInterface/Dropdown.cs
index bdfa9b212..7bcb2eff3 100644
--- a/osu.Framework/Graphics/UserInterface/Dropdown.cs
+++ b/osu.Framework/Graphics/UserInterface/Dropdown.cs
@@ -355,18 +355,12 @@ protected override void OnMouseUp(MouseUpEvent e)
             if (!ToggleOnMouseDown)
                 return;
 
-            // Close dropdown when cursor is released outside the menu
-            if (!Menu.IsHovered)
-            {
-                // Do not close the menu if we are releasing on the DropdownHeader
-                if (!Header.IsHovered)
-                    Menu.Close();
+            if (Header.IsHovered)
                 return;
-            }
 
-            // Cursor is inside the menu and possibly selecting an item,
-            // commit that selection and close the menu
-            ((IDropdown)this).CommitPreselection();
+            if (Menu.IsHovered)
+                ((IDropdown)this).CommitPreselection();
+
             Menu.Close();
         }
 

```

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs (hunk 2 above)
Comment (jarvis394) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: Was looking on 0fed903d87150ac90d4215c718010c00bc164fc4 (https://github.com/ppy/osu-framework/pull/6286); I think we can reintroduce focus methods for `DropdownSearchBar` and keep focus management inside it as well:
```csharp
/// <summary>
/// Focuses the textbox and opens parent <see cref="Dropdown{T}"/> as a result.
/// </summary>
public void ObtainFocus()
{
	dropdown.ChangeFocus(textBox);
}

/// <summary>
/// Clears current search and removes focus from the textbox,
/// closing parent <see cref="Dropdown{T}"/> as a result.
/// </summary>
public void ReleaseFocus()
{
	// Reset states when the menu is closed by any means.
	SearchTerm.Value = string.Empty;

	if (!textBox.HasFocus)
		return;

	dropdown.ChangeFocus(null);
}

/// <summary>
/// Handles changes to the menu visibility.
/// </summary>
private void onMenuStateChanged(MenuState state)
{
	if (state == MenuState.Closed)
		ReleaseFocus();
	else
		ObtainFocus();

	updateTextBoxVisibility();
}
```

Another solution I've tried is to let DropdownSearchBar manage focus on its own and invoke `MenuStateChanged` by closing and opening the menu (it won't invoke if previous value matches current) which is very strange to see in a `dropdown.MenuState == MenuState.Open` if statement. Also it makes textbox lose focus temporarily.

```csharp
private bool onClick(ClickEvent e)
	if (ToggleOnMouseDown) {
		if (dropdown.MenuState == MenuState.Closed) return false

		dropdown.CloseMenu();
		dropdown.OpenMenu();

		return false;
	}
	...
}
```

Obtain/ReleaseFocus works fine and provides public methods for the header, but maybe @smoogipoo has something to say for it.

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs (hunk 1 above)
Comment (jarvis394) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: Your question is towards naming or the setter?

It's easier to setup for tests if it has a setter, same was done above for `AlwaysShowSearchBar`. For real use, it is supposed that it will never be reassigned after init, but nothing will be broken if it is.

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs (hunk 3 above)
Comment (jarvis394) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs:174]: `TestPaddedSearchBar` should do that, it passes with `ToggleOnMouseDown=true` but when I test `ShearedDropdown` with a label in osu!, dropdown closes without this check. I'll try to replicate sheared dropdown structure in a test

File: osu.Framework/Graphics/UserInterface/DropdownHeader.cs (hunk 1 above)
Comment (Joehuu) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: Probably needed to disable on dropdowns inside a scroll container, else you can't drag start scroll on a dropdown anymore.

Comment (bdach) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: my question is towards the mere existence of the flag

> Probably needed to disable on dropdowns inside a scroll container, else you can't drag start scroll on a dropdown anymore.

this sort of thing makes me question the entire premise of this change. are we *really* going to have dropdowns have different UX depending on whether they're inside a scroll or not? that's horrible

Comment (vatei) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: > are we _really_ going to have dropdowns have different UX depending on whether they're inside a scroll or not? that's horrible

How about auto scrolling slowly, one item by one, when you get past the bottom? 
Tried mimicking what I mean:

https://github.com/user-attachments/assets/67f91fa9-d772-40bb-8945-f28c50e38ae9

Comment (Joehuu) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: What I'm talking about is dropdowns in settings. For example, on mobile, you drag to scroll. But if you were to touch a dropdown header before dragging, it won't work.

Comment (jarvis394) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: @bdach,
> are we really going to have dropdowns have different UX depending on whether they're inside a scroll or not? that's horrible

Elements inside scrollable containers **already have** different UX by definition. With a touch, you expect scroll to happen when you drag your finger. When you touch and lift your finger, you expect a click to occur. Fixed surfaces can't scroll, so user's intent will always be a click by default (and then maybe a long-press, etc.) Why don't we skip a step then if we already know user's intent?

You can find these click optimizations in a browser for example:
- on PC tabs list surface is fixed and non-scrollable, therefore tabs switch on mouse down,
- on mobile tabs are inside a scroll container, therefore they open on click.

On the functional side, "power" users will be happy as they can select items quicker (e.g. beatmap collection) while regular ones won't notice the difference (except that menus will open slightly faster). Also, this UX was in stable for years.

Comment (jarvis394) [reply osu.Framework/Graphics/UserInterface/DropdownHeader.cs]: @vatei,
> How about auto scrolling slowly, one item by one, when you get past the bottom?

Probably could be lazily done by adding offset param to `ScrollContainer.ScrollIntoView`, but better solution will be adding an area on top/bottom of the menu that will auto-scroll in a given direction by one item height. It should also have something like a shadow gradient or an arrow to show its interactive. overall, this requires design changes and is just not in the scope of this PR.

[osu.Framework.Tests_DehYRpHzlm.webm](https://github.com/user-attachments/assets/de5eed0a-26ed-45ec-9ead-c9b5634756db)

Collections dropdown max height can also just be bigger :p

**Comment by bdach** [comment]
For clarity I'm not willing to review this further until I get buy-in from @ppy/team-client that everyone is fine with dropdowns changing UX on the fly depending on whether they're inside a scroll or not (https://github.com/ppy/osu-framework/pull/6610#discussion_r2199668728).

**Comment by peppy** [comment]
> For clarity I'm not willing to review this further until I get buy-in from @ppy/team-client that everyone is fine with dropdowns changing UX on the fly depending on whether they're inside a scroll or not ([#6610 (comment)](https://github.com/ppy/osu-framework/pull/6610#discussion_r2199668728)).

I'm fine with this as a UX behaviour. As for implementation, we may want to consider making it automatic via inferring a parent scroll container?

https://github.com/ppy/osu-framework/blob/81663545601e0f9dcea461223163354ce2f3c0ca/osu.Framework/Graphics/Containers/ScrollContainer.cs#L193-L196

**Comment by jarvis394** [comment]
> As for implementation, we may want to consider making it automatic via inferring a parent scroll container?

https://github.com/user-attachments/assets/5ac0904d-2f72-42b9-ac39-4d2f7c9817b8

Looks like `this.FindClosestParent<IScrollContainer>()` is doing wonders! No changes outside the framework are needed.

I've checked most of the places where dropdown menus are used in osu! and all look great, although editor's top bar is scrollable and its menus will therefore open on click – I guess it's fine anyways. Added a test with dropdown in a scrollable parent as well.

The only edge case for this behaviour is when dropdown header is moved somewhere from cursor. This is the unwanted UX change bdach was concerned about. In my defense, I would say this is rare (not present anywhere in osu! at least) and I would consider fixing the layout or just implicitly setting `Dropdown.ToggleOnMouseDown` to `false`.

https://github.com/user-attachments/assets/82da633c-e210-4e20-8b48-bf0eb16a297d
